<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8" />
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="viewport" content="width=device-width, initial-scale=1">

<meta name="author" content="Paul Staab" />


<title>Using External Simulators</title>



<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; color: #aaaaaa; border-right: 1px solid #aaaaaa; }
td.sourceCode { padding-left: 5px; }
code > span.kw { color: #007020; font-weight: bold; } /* Keyword */
code > span.dt { color: #902000; } /* DataType */
code > span.dv { color: #40a070; } /* DecVal */
code > span.bn { color: #40a070; } /* BaseN */
code > span.fl { color: #40a070; } /* Float */
code > span.ch { color: #4070a0; } /* Char */
code > span.st { color: #4070a0; } /* String */
code > span.co { color: #60a0b0; font-style: italic; } /* Comment */
code > span.ot { color: #007020; } /* Other */
code > span.al { color: #ff0000; font-weight: bold; } /* Alert */
code > span.fu { color: #06287e; } /* Function */
code > span.er { color: #ff0000; font-weight: bold; } /* Error */
code > span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
code > span.cn { color: #880000; } /* Constant */
code > span.sc { color: #4070a0; } /* SpecialChar */
code > span.vs { color: #4070a0; } /* VerbatimString */
code > span.ss { color: #bb6688; } /* SpecialString */
code > span.im { } /* Import */
code > span.va { color: #19177c; } /* Variable */
code > span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code > span.op { color: #666666; } /* Operator */
code > span.bu { } /* BuiltIn */
code > span.ex { } /* Extension */
code > span.pp { color: #bc7a00; } /* Preprocessor */
code > span.at { color: #7d9029; } /* Attribute */
code > span.do { color: #ba2121; font-style: italic; } /* Documentation */
code > span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code > span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code > span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
</style>



<link href="data:text/css;charset=utf-8,body%20%7B%0Abackground%2Dcolor%3A%20%23fff%3B%0Amargin%3A%201em%20auto%3B%0Amax%2Dwidth%3A%20700px%3B%0Aoverflow%3A%20visible%3B%0Apadding%2Dleft%3A%202em%3B%0Apadding%2Dright%3A%202em%3B%0Afont%2Dfamily%3A%20%22Open%20Sans%22%2C%20%22Helvetica%20Neue%22%2C%20Helvetica%2C%20Arial%2C%20sans%2Dserif%3B%0Afont%2Dsize%3A%2014px%3B%0Aline%2Dheight%3A%201%2E35%3B%0A%7D%0A%23header%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0A%23TOC%20%7B%0Aclear%3A%20both%3B%0Amargin%3A%200%200%2010px%2010px%3B%0Apadding%3A%204px%3B%0Awidth%3A%20400px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Aborder%2Dradius%3A%205px%3B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Afont%2Dsize%3A%2013px%3B%0Aline%2Dheight%3A%201%2E3%3B%0A%7D%0A%23TOC%20%2Etoctitle%20%7B%0Afont%2Dweight%3A%20bold%3B%0Afont%2Dsize%3A%2015px%3B%0Amargin%2Dleft%3A%205px%3B%0A%7D%0A%23TOC%20ul%20%7B%0Apadding%2Dleft%3A%2040px%3B%0Amargin%2Dleft%3A%20%2D1%2E5em%3B%0Amargin%2Dtop%3A%205px%3B%0Amargin%2Dbottom%3A%205px%3B%0A%7D%0A%23TOC%20ul%20ul%20%7B%0Amargin%2Dleft%3A%20%2D2em%3B%0A%7D%0A%23TOC%20li%20%7B%0Aline%2Dheight%3A%2016px%3B%0A%7D%0Atable%20%7B%0Amargin%3A%201em%20auto%3B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dcolor%3A%20%23DDDDDD%3B%0Aborder%2Dstyle%3A%20outset%3B%0Aborder%2Dcollapse%3A%20collapse%3B%0A%7D%0Atable%20th%20%7B%0Aborder%2Dwidth%3A%202px%3B%0Apadding%3A%205px%3B%0Aborder%2Dstyle%3A%20inset%3B%0A%7D%0Atable%20td%20%7B%0Aborder%2Dwidth%3A%201px%3B%0Aborder%2Dstyle%3A%20inset%3B%0Aline%2Dheight%3A%2018px%3B%0Apadding%3A%205px%205px%3B%0A%7D%0Atable%2C%20table%20th%2C%20table%20td%20%7B%0Aborder%2Dleft%2Dstyle%3A%20none%3B%0Aborder%2Dright%2Dstyle%3A%20none%3B%0A%7D%0Atable%20thead%2C%20table%20tr%2Eeven%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Ap%20%7B%0Amargin%3A%200%2E5em%200%3B%0A%7D%0Ablockquote%20%7B%0Abackground%2Dcolor%3A%20%23f6f6f6%3B%0Apadding%3A%200%2E25em%200%2E75em%3B%0A%7D%0Ahr%20%7B%0Aborder%2Dstyle%3A%20solid%3B%0Aborder%3A%20none%3B%0Aborder%2Dtop%3A%201px%20solid%20%23777%3B%0Amargin%3A%2028px%200%3B%0A%7D%0Adl%20%7B%0Amargin%2Dleft%3A%200%3B%0A%7D%0Adl%20dd%20%7B%0Amargin%2Dbottom%3A%2013px%3B%0Amargin%2Dleft%3A%2013px%3B%0A%7D%0Adl%20dt%20%7B%0Afont%2Dweight%3A%20bold%3B%0A%7D%0Aul%20%7B%0Amargin%2Dtop%3A%200%3B%0A%7D%0Aul%20li%20%7B%0Alist%2Dstyle%3A%20circle%20outside%3B%0A%7D%0Aul%20ul%20%7B%0Amargin%2Dbottom%3A%200%3B%0A%7D%0Apre%2C%20code%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0Aborder%2Dradius%3A%203px%3B%0Acolor%3A%20%23333%3B%0Awhite%2Dspace%3A%20pre%2Dwrap%3B%20%0A%7D%0Apre%20%7B%0Aborder%2Dradius%3A%203px%3B%0Amargin%3A%205px%200px%2010px%200px%3B%0Apadding%3A%2010px%3B%0A%7D%0Apre%3Anot%28%5Bclass%5D%29%20%7B%0Abackground%2Dcolor%3A%20%23f7f7f7%3B%0A%7D%0Acode%20%7B%0Afont%2Dfamily%3A%20Consolas%2C%20Monaco%2C%20%27Courier%20New%27%2C%20monospace%3B%0Afont%2Dsize%3A%2085%25%3B%0A%7D%0Ap%20%3E%20code%2C%20li%20%3E%20code%20%7B%0Apadding%3A%202px%200px%3B%0A%7D%0Adiv%2Efigure%20%7B%0Atext%2Dalign%3A%20center%3B%0A%7D%0Aimg%20%7B%0Abackground%2Dcolor%3A%20%23FFFFFF%3B%0Apadding%3A%202px%3B%0Aborder%3A%201px%20solid%20%23DDDDDD%3B%0Aborder%2Dradius%3A%203px%3B%0Aborder%3A%201px%20solid%20%23CCCCCC%3B%0Amargin%3A%200%205px%3B%0A%7D%0Ah1%20%7B%0Amargin%2Dtop%3A%200%3B%0Afont%2Dsize%3A%2035px%3B%0Aline%2Dheight%3A%2040px%3B%0A%7D%0Ah2%20%7B%0Aborder%2Dbottom%3A%204px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Apadding%2Dbottom%3A%202px%3B%0Afont%2Dsize%3A%20145%25%3B%0A%7D%0Ah3%20%7B%0Aborder%2Dbottom%3A%202px%20solid%20%23f7f7f7%3B%0Apadding%2Dtop%3A%2010px%3B%0Afont%2Dsize%3A%20120%25%3B%0A%7D%0Ah4%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23f7f7f7%3B%0Amargin%2Dleft%3A%208px%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Ah5%2C%20h6%20%7B%0Aborder%2Dbottom%3A%201px%20solid%20%23ccc%3B%0Afont%2Dsize%3A%20105%25%3B%0A%7D%0Aa%20%7B%0Acolor%3A%20%230033dd%3B%0Atext%2Ddecoration%3A%20none%3B%0A%7D%0Aa%3Ahover%20%7B%0Acolor%3A%20%236666ff%3B%20%7D%0Aa%3Avisited%20%7B%0Acolor%3A%20%23800080%3B%20%7D%0Aa%3Avisited%3Ahover%20%7B%0Acolor%3A%20%23BB00BB%3B%20%7D%0Aa%5Bhref%5E%3D%22http%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0Aa%5Bhref%5E%3D%22https%3A%22%5D%20%7B%0Atext%2Ddecoration%3A%20underline%3B%20%7D%0A%0Acode%20%3E%20span%2Ekw%20%7B%20color%3A%20%23555%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Edt%20%7B%20color%3A%20%23902000%3B%20%7D%20%0Acode%20%3E%20span%2Edv%20%7B%20color%3A%20%2340a070%3B%20%7D%20%0Acode%20%3E%20span%2Ebn%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Efl%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Ech%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Est%20%7B%20color%3A%20%23d14%3B%20%7D%20%0Acode%20%3E%20span%2Eco%20%7B%20color%3A%20%23888888%3B%20font%2Dstyle%3A%20italic%3B%20%7D%20%0Acode%20%3E%20span%2Eot%20%7B%20color%3A%20%23007020%3B%20%7D%20%0Acode%20%3E%20span%2Eal%20%7B%20color%3A%20%23ff0000%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%0Acode%20%3E%20span%2Efu%20%7B%20color%3A%20%23900%3B%20font%2Dweight%3A%20bold%3B%20%7D%20%20code%20%3E%20span%2Eer%20%7B%20color%3A%20%23a61717%3B%20background%2Dcolor%3A%20%23e3d2d2%3B%20%7D%20%0A" rel="stylesheet" type="text/css" />

</head>

<body>




<h1 class="title toc-ignore">Using External Simulators</h1>
<h4 class="author"><em>Paul Staab</em></h4>
<h4 class="date"><em>coala 0.5.2</em></h4>



<p>Coala can call the coalescent simulators <em>ms</em>[1], <em>msms</em>[2] and <em>scrm</em>[3] and can use <em>seq-gen</em>[4] for finite sites simulations. The R version of <em>scrm</em> should get installed automatically as a dependeny of coala. For the other programs, you need to have an executable binary available on your system.</p>
<div id="installation" class="section level1">
<h1>Installation</h1>
<p>Short instructions on obtaining and compiling the programs are given in the help pages of <code>activate_ms</code>, <code>activate_msms</code> and <code>activate_seqgen</code>. More detailed instructions are provided in <a href="https://github.com/statgenlmu/coala/wiki/Installation">the wiki</a>.</p>
</div>
<div id="activation" class="section level1">
<h1>Activation</h1>
<p>In addition to providing the binary for a simulator, you need to inform <code>coala</code> where the binary is. We refer to this process as <em>activation</em> of a binary. Afterwards, <code>coala</code> will use the simulator automatically where-ever appropriate.</p>
<p>There are three different ways to activate a binary:</p>
<ol style="list-style-type: decimal">
<li>Use the <code>activate_msms</code> and <code>activate_seqgen</code> functions to activate the simulators from within R. You should use the functions before creating a model.</li>
<li>Alternatively, you can place the binaries in your working directory or in a folder listed in your <em>PATH</em> environment variable using one of the names listed under “Expected Binary Names” below. If there is a match file, coala will automatically activate the simulator.</li>
<li>You can start the R session with an environment variable that hold the path to the binaries. In this case, the simulators should also be automatically be activated when the coala package is loaded.</li>
<li>Coala uses the R versions of <code>scrm</code> and <code>ms</code>. <code>scrm</code> should alawys be available. Install the CRAN package <code>phyclust</code> to use <code>ms</code>.</li>
</ol>
<table>
<thead>
<tr class="header">
<th>Simulator</th>
<th>Priority</th>
<th>Expected Binary Names</th>
<th>Environment Var</th>
<th>Function</th>
</tr>
</thead>
<tbody>
<tr class="odd">
<td>seq-gen</td>
<td>100</td>
<td>seqgen, seq-gen, seqgen.exe, seq-gen.exe</td>
<td>SEQGEN</td>
<td>activate_seqgen</td>
</tr>
<tr class="even">
<td>msms</td>
<td>200</td>
<td>msms.jar / java, java.exe</td>
<td>MSMS / JAVA</td>
<td>activate_msms</td>
</tr>
<tr class="odd">
<td>ms</td>
<td>300</td>
<td></td>
<td></td>
<td>activate_ms</td>
</tr>
<tr class="even">
<td>scrm</td>
<td>400</td>
<td></td>
<td></td>
<td></td>
</tr>
</tbody>
</table>
<p>You can use the <code>list_simulators()</code> command to view which simulators are currently available:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(coala)
<span class="kw">list_simulators</span>()</code></pre></div>
<pre><code>##   name priority                      info
## 2 scrm      400         version : 1.7.2-0
## 1   ms      300 version : phyclust_0.1.20</code></pre>
</div>
<div id="priority" class="section level1">
<h1>Priority</h1>
<p>The <code>check_model</code> function checks which simulators support a specific model, and states the problems which coala has detected with the simulators that do not support it. For example, a simple model with infinite-sites mutations (IFS) can be simulated with <code>scrm</code> or – if installed – with <code>ms</code> and <code>msms</code>, but not with <code>seq-gen</code> because the latter generates finite-sites mutations:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">model &lt;-<span class="st"> </span><span class="kw">coal_model</span>(<span class="dv">10</span>, <span class="dv">1</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">feat_mutation</span>(<span class="dv">5</span>, <span class="dt">model =</span> <span class="st">&quot;IFS&quot;</span>) <span class="op">+</span>
<span class="st">  </span><span class="kw">sumstat_nucleotide_div</span>()
<span class="kw">check_model</span>(model)</code></pre></div>
<pre><code>## ms : OK
## 
## scrm : OK</code></pre>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r">model</code></pre></div>
<pre><code>## Features: 
## * Sampling of 10 (pop 1) individuals with ploidy 1 at time `0` 
## * Mutations with rate `5` following a IFS mutation model
## * Generating Seg. Sites
## 
## Parameter: None
## 
## Loci: 1 locus of length 1000 
## 
## Summary Statistics: stat_pi 
## 
## Simulator: scrm 
## Command: scrm 10 1 -t 5</code></pre>
<p>If multiple simulators can simulate a model, the one with the highest priority is used. In our example, that is <code>scrm</code>. If we would like to use <code>ms</code> instead, we need to raise its priority:</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">activate_ms</span>(<span class="dt">priority =</span> <span class="dv">500</span>)</code></pre></div>
</div>
<div id="references" class="section level1">
<h1>References</h1>
<ul>
<li><strong>[1]</strong>: Richard R. Hudson. <em>Generating samples under a Wright-Fisher neutral model of genetic variation.</em> Bioinformatics (2002) 18 (2): 337-338 <a href="http://doi.org/10.1093/bioinformatics/18.2.337">10.1093/bioinformatics/18.2.337</a></li>
<li><strong>[2]</strong>: Gregory Ewing and Joachim Hermisson. <em>MSMS: a coalescent simulation program including recombination, demographic structure and selection at a single locus.</em> Bioinformatics (2010) 26 (16): 2064-2065 <a href="http://doi.org/10.1093/bioinformatics/btq322">10.1093/bioinformatics/btq322</a></li>
<li><strong>[3]</strong>: Paul R. Staab, Sha Zhu, Dirk Metzler and Gerton Lunter. <em>scrm: efficiently simulating long sequences using the approximated coalescent with recombination.</em> Bioinformatics (2015) 31 (10): 1680-1682. <a href="http://doi.org/doi:10.1093/bioinformatics/btu861">10.1093/bioinformatics/btu861</a></li>
<li><strong>[4]</strong>: Andrew Rambaut and Nicholas C. Grassly. <em>Seq-Gen: an application for the Monte Carlo simulation of DNA sequence evolution along phylogenetic trees.</em> Comput Appl Biosci (1997) 13 (3): 235-238 <a href="http://doi.org/10.1093/bioinformatics/13.3.235">10.1093/bioinformatics/13.3.235</a></li>
</ul>
</div>



<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
